src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,175 @@
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract partial class AbstractTypeImportCompletionProvider : CommonCompletionProvider
+    {
+        protected abstract Task<SyntaxContext> CreateContextAsync(Document document, int position, CancellationToken cancellationToken);
+
+        protected abstract void GetImportedNamespaces(
+            SyntaxNode location,
+            SemanticModel semanticModel,
+            ImmutableHashSet<string>.Builder builder,
+            CancellationToken cancellationToken);
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+            var document = completionContext.Document;
+            var cancellationToken = completionContext.CancellationToken;
+            var workspace = document.Project.Solution.Workspace;
+            var experimentationService = workspace.Services.GetService<IExperimentationService>();
+
+            var importCompletionOptionValue = completionContext.Options.GetOption(CompletionOptions.ShowImportCompletionItems, document.Project.Language);
+
+            // Don't trigger import completion if the option value is "default" and the experiment is disabled for the user. 
+            if (importCompletionOptionValue == false ||
+                (importCompletionOptionValue == null && experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.TypeImportCompletion) != true))
+            {
+                return;
+            }
+
+            var syntaxContext = await CreateContextAsync(document, completionContext.Position, cancellationToken).ConfigureAwait(false);
+            if (!syntaxContext.IsTypeContext)
+            {
+                return;
+            }
+
+            using (Logger.LogBlock(FunctionId.Completion_TypeImportCompletionProvider_GetCompletionItemsAsync, cancellationToken))
+            {
+                var project = document.Project;
+                var typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+
+                // Find all namespaces in scope at current cursor location, 
+                // which will be used to filter so the provider only returns out-of-scope types.
+                var namespacesInScope = GetNamespacesInScope(document, syntaxContext, cancellationToken);
+                Action<CompletionItem> handleAccessibleItem = item => AddItems(item, completionContext, namespacesInScope);
+
+                // Get completion items from current project.
+                await typeImportCompletionService.GetAccessibleTopLevelTypesFromProjectAsync(project, handleAccessibleItem, cancellationToken)
```

Comment:
is there a particular reason why the provider does the namespaceInSocpe and addItem rather than typeImportCompletionService does it? it feels it would be hard to use the service outside of this provider.

from syntaxContext, all it uses is location and position which can just given to GetAccessibleTopLevelTypesFromProjectAsync

Reply:
or GetAccessibleTopLevelTypesFromProjectAsync could just return all items and the caller do the filtering. the accessibility filter seems done outside of cache anyway.

Reply:
Fixed, now provider handles all the filtering.